Send confirmation SMS on registration

brucebolt opened the pull request:
What
----

Sends a confirmation text message to registrants who have entered a UK mobile telephone number into the "phone number for text messages" field.

Content is currently awaiting departmental sign-off, so this PR cannot yet be merged.

Example of SMS:
![kAJUmrd8](https://user-images.githubusercontent.com/6329861/81176726-768fe280-8f9d-11ea-91c0-2783d83559bb.png)

How to review
-------------

- Review code changes.
- Set up and run application locally using a Notify API token.  Ensure text message is sent through Notify interface.

Links
-----

Trello card: https://trello.com/c/8L0b1iIg

brucebolt committed:
Make it clear Notify template ID is for emails

We will soon be sending SMS confirmation, which has a different Notify
template ID.  Renaming the existing environment variable so we are clear
this relates to emails.

brucebolt committed:
Add Notify SMS template IDs

brucebolt committed:
Clarify Notify Delivery Method only sends emails

We will soon be sending SMS confirmations, so need to clarify that the
existing delivery method only sends emails.

brucebolt committed:
Add SMS mailer for Notify

This mailer will allow us to send text messages, in addition to the
existing support for email confirmations.

brucebolt committed:
Add confirmation SMS content to locales

brucebolt committed:
Update README to include information about sending SMSs

brucebolt committed:
Send a SMS confirmation on registration

Not all users enter an email address, and some enter an invalid email
address.  We are therefore sending SMS confirmation using GOV.UK Notify
to ensure these users are aware their registration is complete.

bilbof commented on README.md:
```diff
@@ -51,17 +51,18 @@ to process the email queue.
-You'll also need to set a `GOVUK_NOTIFY_TEMPLATE_ID`, which might involve
-creating a template in Notify if your Notify service doesn't have one.
+You'll also need to set a `GOVUK_NOTIFY_EMAIL_TEMPLATE_ID` and
+`GOVUK_NOTIFY_SMS_TEMPLATE_ID`, which might involvei creating a
```
```suggestion
`GOVUK_NOTIFY_SMS_TEMPLATE_ID`, which might involve creating a
```

bilbof commented on app/controllers/coronavirus_form/check_answers_controller.rb:
```diff
@@ -27,6 +27,7 @@ def submit
 
       send_confirmation_email if session_with_indifferent_access.dig(:contact_details, :email).present?
+      send_confirmation_sms if session_with_indifferent_access.dig(:contact_details, :phone_number_texts).present? && mobile_number_provided?
     end
```
I wonder if we also want to update the `smoke_tester?` method to check if the mobile number used is one used by the smoke tester?

brucebolt replied:
Good idea.  I've added a condition for this.

bilbof commented on app/controllers/coronavirus_form/check_answers_controller.rb:
```diff
@@ -64,6 +74,14 @@ def user_email
+  end
+
+  def mobile_number_provided?
+    TelephoneNumber.valid?(session_with_indifferent_access.dig(:contact_details, :phone_number_texts), :gb, [:mobile])
```
We'll need to remove whitespace from the mobile number here, if this isn't done elsewhere.

brucebolt replied:
I think we will have done this when the number is stored in the contact details controller, since the number is being saved as `national_number` which will have already been formatted by the TelephoneNumber gem.

bilbof commented on concourse/tasks/deploy-to-govuk-paas.yml:
```diff
@@ -60,7 +61,8 @@ run:
       cf set-env govuk-coronavirus-vulnerable-people-form METRICS_PASSWORD "$METRICS_PASSWORD"
-      cf set-env govuk-coronavirus-vulnerable-people-form GOVUK_NOTIFY_TEMPLATE_ID "$GOVUK_NOTIFY_TEMPLATE_ID"
+      cf set-env govuk-coronavirus-vulnerable-people-form GOVUK_NOTIFY_EMAIL_TEMPLATE_ID "$GOVUK_NOTIFY_EMAIL_TEMPLATE_ID"
+      cf set-env govuk-coronavirus-vulnerable-people-form GOVUK_NOTIFY_SMS_TEMPLATE_ID "$GOVUK_NOTIFY_SMS_TEMPLATE_ID"
```
`GOVUK_NOTIFY_TEMPLATE_ID` will still be set on the app, so we might want to unset that at some point once this is deployed.

brucebolt replied:
Cool, I'll do this manually after deployment of these changes

bilbof commented on config/locales/en.yml:
```diff
@@ -191,6 +191,20 @@ en:
+  sms:
+    confirmation:
+      body_text: |
+        Hello %{first_name} %{last_name} - we’ve received your registration.
```
Do we want to limit the length of `first_name` or `last_name`, and also remove any special characters (e.g. `.`). Spammers could use this to send texts to people with e.g. some spam message, perhaps containing a link. This isn't specific to the SMS though..

brucebolt replied:
I'm hesitant to do this, since some names can be rather long and include special characters [1].  We have around 300 characters remaining in the SMS, so I don't think length is an issue.

Maybe we should have another story to perform some validation on these fields when the data is actually entered so anything that looks like a URL is disallowed, thereby preventing phishing attempts using this form?

[1]: https://www.bbc.co.uk/news/world-us-canada-52557291

bilbof commented on spec/controllers/coronavirus_form/check_answers_controller_spec.rb:
```diff
@@ -72,6 +72,29 @@
+        .with(a_hash_including(reference_number: "abc"), "07790900000").on_queue("mailers")
+    end
+
+    it "does not send a text message when no email address provided" do
```
```suggestion
    it "does not send a text message when no phone number for texting is provided" do
```

bilbof reviewed (commented):
Looks great, just a few comments.

It sounds like there are other things to consider (e.g. text content) too, so will approve when those are done.

brucebolt committed:
Do not send text messages when running tests

We already do not send emails when running tests (either locally or as
smoke tests).  It is therefore prudent to do the same for SMS.

brucebolt commented:
I've pushed an additional commit that will override the sender from "GOVUK" to the outgoing number for the National Shielding Service call centre.  Also configured this number in the Notify dashboard.

brucebolt commented:
This has been rebased against master and tested on staging.

I am awaiting confirmation on the sender name before we can merge and deploy to production.

brucebolt commented:
We've had confirmation the sender should be `GOVUK` (i.e. the default).  I have therefore removed the final commit for changing the sender.

bilbof commented on config/locales/en.yml:
```diff
@@ -192,6 +192,20 @@ en:
+
+        If you need urgent support and can’t rely on friends or neighbours, contact your local authority.
+
+        If your support needs change, go to https://www.gov.uk/coronavirus-extremely-vulnerable and answer the questions again.
```
Do you know if this'll be one text based on the length? Just had a thought we might want a test for the length of this content, just in case.

brucebolt replied:
It will cost multiple text messages to send (5x SMSs IIRC), but it will be delivered as a single message to the recipient's device

bilbof replied:
Wow big. That's good to know 👍

bilbof reviewed (approved):
Excellent work 👍
